### PR TITLE
PR: Add option to toggle Emacs keybindings

### LINKED ISSUE #669: Provide Vim and Emacs key bindings for the editor
_From [r.c.bruno.andre@gmail.com](https://code.google.com/u/r.c.bruno.andre@gmail.com/) on 2011-05-27T12:35:31Z_

Could it be possible to use vim as the text editor in Spyder?

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=669_

### PR DESCRIPTION
**Pull request by blink1073**
The keyboard shortcuts themselves are open for discussion.  I did my best to match this [table](http://wttools.sourceforge.net/emacs-stuff/emacs-keybindings.html).

CAM-Gerlach Edit: Fixes #669 

screen shot 2015-09-07 at 10 27 46 am

screen shot 2015-09-07 at 10 27 54 am

### CODE REVIEW

**Comment by goanpeca** [comment]
@Nodd

**Comment by Nodd** [comment]
Are the emacs keybindings read-only ? In this case, checking the button should disable the edition. Also the checkbox should be on the left side, with "Emacs Keybindings as its label.

If it's not, I'd prefer a pushbutton with a label like "Load emacs-like keybondings". If we add more layouts later, the pushbutton could be a QListWidget. I wouldn't expect a checkbox here.

**Comment by ccordoba12** [comment]
I'd like to have a combobox instead of a checkbox, so that we can add also IPython and Vim keybindings later on :-)

**Comment by blink1073** [comment]
Updated.  The old reset behavior is now achieved by explicitly loading the defaults.

**Comment by goanpeca** [comment]
@blink1073 what do you mean by
>  I did my best to match this

Some are not matching?

Also we need some way to handle custom shortcuts? maybe IO load the defaults and change a couple and if for some reason I  load VIM bindings then I lost the changes I made to the spyder defaults... 

Otherwise I think it is pretty cool :-)

**Comment by blink1073** [comment]
@goanpeca, I mean, Emacs uses different terminology and I did my best to translate (I only used Emacs for like 3 months a few years ago).

Yeah, it would be nice to store a set of custom shortcuts, but that is outside the scope of this PR.  :wink:

**Comment by goanpeca** [comment]
Ahhh ok got it :-) 

Sure, not for this PR, but it is good that we agree that it might be useful, cheers

**Comment by ccordoba12** [comment]
Thanks a lot @blink1073!! This is looking really good!

I hope to merge it soon :-)

File: spyderlib/config/main.py [hunk 1]
```diff
@@ -580,6 +580,53 @@ def is_ubuntu():
...
+              'editor/redo': 'Ctrl+X, Ctrl+Y',
+              'editor/cut': 'Ctrl+W',
+              'editor/copy': 'Ctrl+C',
+              'editor/paste': 'Ctrl+Y',
+              'editor/delete': 'Delete',
+              'editor/select all': "Ctrl+X+H",
+              # -- In plugins/editor
+              'editor/new file': "Ctrl+X, Ctrl+N",
+              'editor/open file': "Ctrl+X, Ctrl+F",
+              'editor/save file': "Ctrl+X, Ctrl+W",
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: This one is wrong, it should be `Ctrl+X, Ctrl+S`

File: spyderlib/config/main.py [hunk 2]
```diff
@@ -580,6 +580,53 @@ def is_ubuntu():
...
+              'editor/save all': "Ctrl+Alt+S",
+              'editor/save as': 'Ctrl+Shift+S',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: And this one is `Ctrl+X, Ctrl+W`

File: spyderlib/config/main.py [hunk 3]
```diff
@@ -580,6 +580,53 @@ def is_ubuntu():
...
+              'editor/go to line': 'Ctrl+X, Ctrl+L',
+              'editor/go to previous file': 'Ctrl+Tab',
+              'editor/go to next file': 'Ctrl+Shift+Tab',
+              '_/file switcher': 'Meta+Shift+P',
+              # -- In spyder.py
+              'editor/find text': "Ctrl+S",
+              'editor/find next': "F3",
+              'editor/find previous': "Ctrl+R",
+              'editor/replace text': "Ctrl+X, Ctrl+H",
+              'editor/undo': 'Meta+Backspace',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: This is `Ctrl+_`

File: spyderlib/config/main.py [hunk 4]
```diff
@@ -580,6 +580,53 @@ def is_ubuntu():
...
+              'editor/end of document': 'Meta+Right',
+              # -- In widgets/editor
+              'editor/inspect current object': 'Ctrl+X, Ctrl+I',
+              'editor/go to line': 'Ctrl+X, Ctrl+L',
+              'editor/go to previous file': 'Ctrl+Tab',
+              'editor/go to next file': 'Ctrl+Shift+Tab',
+              '_/file switcher': 'Meta+Shift+P',
+              # -- In spyder.py
+              'editor/find text': "Ctrl+S",
+              'editor/find next': "F3",
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: This one is also `Ctrl+S`. I don't know what to do about it :-)

File: spyderlib/config/main.py [hunk 5]
```diff
@@ -580,6 +580,53 @@ def is_ubuntu():
               # ---- Profiler (in p_profiler) ----
               'profiler/run profiler': "F10"
               }),
+            ('emacs_shortcuts',
+              {
+              # -- In codeeditor
+              'editor/code completion': 'Meta+Tab',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: I think we should leave `Ctrl+Space` here because Emacs doesn't have the notion of completion.

File: spyderlib/config/main.py [hunk 6]
```diff
@@ -580,6 +580,53 @@ def is_ubuntu():
...
+              # -- In codeeditor
+              'editor/code completion': 'Meta+Tab',
+              'editor/go to definition': "Ctrl+G",
+              'editor/start of line': "Ctrl+A",
+              'editor/end of line': "Ctrl+E",
+              'editor/previous line': "Ctrl+P",
+              'editor/next line': "Ctrl+N",
+              'editor/previous char': "Ctrl+B",
+              'editor/next char': "Ctrl+F",
+              'editor/previous word': "Meta+B",
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: This one is `Alt+B`

File: spyderlib/config/main.py [hunk 7]
```diff
@@ -580,6 +580,53 @@ def is_ubuntu():
...
+              'editor/next word': "Meta+F",
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: And this one is `Alf+F`

File: spyderlib/config/main.py [hunk 8]
```diff
@@ -580,6 +580,53 @@ def is_ubuntu():
...
+              'editor/kill to line end': "Ctrl+K",
+              'editor/kill to line start': "Ctrl+U",
+              'editor/yank': 'Meta+Y',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: `yank` is equivalent to `paste`, right?

File: spyderlib/config/main.py [hunk 9]
```diff
@@ -580,6 +580,53 @@ def is_ubuntu():
...
+              'editor/rotate kill ring': 'Shift+Ctrl+Y',
+              'editor/kill previous word': 'Ctrl+Backspace',
+              'editor/kill next word': 'Ctrl+Delete',
+              'editor/start of document': 'Meta+Left',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: This is `Alt+<`

File: spyderlib/config/main.py [hunk 10]
```diff
@@ -580,6 +580,53 @@ def is_ubuntu():
...
+              'editor/end of document': 'Meta+Right',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: This is `Alt+>`

File: spyderlib/config/main.py [hunk 11]
```diff
@@ -580,6 +580,53 @@ def is_ubuntu():
...
+              # -- In widgets/editor
+              'editor/inspect current object': 'Ctrl+X, Ctrl+I',
+              'editor/go to line': 'Ctrl+X, Ctrl+L',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: This is `Alt+G, Alt+G`

File: spyderlib/config/main.py [hunk 12]
```diff
@@ -580,6 +580,53 @@ def is_ubuntu():
...
+              'editor/go to previous file': 'Ctrl+Tab',
+              'editor/go to next file': 'Ctrl+Shift+Tab',
+              '_/file switcher': 'Meta+Shift+P',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: This is `Ctrl+X, Ctrl+B`

File: spyderlib/config/main.py [hunk 13]
```diff
@@ -580,6 +580,53 @@ def is_ubuntu():
...
+              # -- In spyder.py
+              'editor/find text': "Ctrl+S",
+              'editor/find next': "F3",
+              'editor/find previous': "Ctrl+R",
+              'editor/replace text': "Ctrl+X, Ctrl+H",
+              'editor/undo': 'Meta+Backspace',
+              'editor/redo': 'Ctrl+X, Ctrl+Y',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: Emacs doesn't have redos ;-)

File: spyderlib/config/main.py [hunk 14]
```diff
@@ -580,6 +580,53 @@ def is_ubuntu():
...
+              'editor/cut': 'Ctrl+W',
+              'editor/copy': 'Ctrl+C',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: This is `Alt+W`

File: spyderlib/config/main.py [hunk 15]
```diff
@@ -580,6 +580,53 @@ def is_ubuntu():
...
+              'editor/paste': 'Ctrl+Y',
+              'editor/delete': 'Delete',
+              'editor/select all': "Ctrl+X+H",
+              # -- In plugins/editor
+              'editor/new file': "Ctrl+X, Ctrl+N",
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: Let's leave this one in `Ctrl+N` because Emacs doesn't have a shortcut for new files

File: spyderlib/config/main.py [hunk 16]
```diff
@@ -580,6 +580,53 @@ def is_ubuntu():
...
+              'editor/open file': "Ctrl+X, Ctrl+F",
+              'editor/save file': "Ctrl+X, Ctrl+W",
+              'editor/save all': "Ctrl+Alt+S",
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: This is `Ctrl+X+S`

File: spyderlib/plugins/shortcuts.py [hunk 17]
```diff
@@ -807,6 +816,22 @@ def apply_settings(self, options):
         self.table.save_shortcuts()
         self.main.apply_shortcuts()
 
+    def load_button_clicked(self, change):
+        selected = self.combo_keybindings.currentText().lower()
+        reset_shortcuts()
```
Comment (ccordoba12) [inline spyderlib/plugins/shortcuts.py]: Do we really need to do this? What would happen with user defined shortcuts in this case?

Comment (goanpeca) [inline spyderlib/plugins/shortcuts.py]: They disappear :-p

File: spyderlib/config/main.py (hunk 11 above)
Comment (ccordoba12) [inline spyderlib/config/main.py]: Sorry, it is `Alt+G+G`

File: spyderlib/plugins/shortcuts.py (hunk 17 above)
Comment (ccordoba12) [inline spyderlib/plugins/shortcuts.py]: Bad thing! We need to find a better solution :-)

Comment (goanpeca) [inline spyderlib/plugins/shortcuts.py]: Yes that is what I told @blink1073, as soon as the user changes the defaults, a `custom` option should appear in the combobox and those shortcuts should be saved for later reuse...

Comment (ccordoba12) [inline spyderlib/plugins/shortcuts.py]: Ok, let's leave this for another PR, as @blink1073 mentioned.

However, I don't know if a `custom` option is enough. What if you want the Emacs shortcuts but do some customizations on top of them? And also for the Default ones?

Comment (goanpeca) [inline spyderlib/plugins/shortcuts.py]: Hmm we would be entering a rabbit hole here...

Comment (ccordoba12) [inline spyderlib/plugins/shortcuts.py]: I think we should preserve customizations for any scheme, and leave the button called `Reset to default schemes` :-)

Comment (goanpeca) [inline spyderlib/plugins/shortcuts.py]: Hmmm sure why not :-9, but for another PR :p

File: spyderlib/config/main.py (hunk 8 above)
Comment (blink1073) [inline spyderlib/config/main.py]: Yes, but it also affects the Kill Ring.

File: spyderlib/config/main.py (hunk 15 above)
Comment (blink1073) [inline spyderlib/config/main.py]: But `Ctrl+N` is go to next line.

**Comment by blink1073** [comment]
Remapped the emacs keybindings.

File: spyderlib/config/main.py [hunk 18]
```diff
@@ -580,6 +580,51 @@ def is_ubuntu():
...
+              'editor/end of line': "Ctrl+E",
+              'editor/previous line': "Ctrl+P",
+              'editor/next line': "Ctrl+N",
+              'editor/previous char': "Ctrl+B",
+              'editor/next char': "Ctrl+F",
+              'editor/previous word': "Alt+B",
+              'editor/next word': "Alt+F",
+              'editor/kill to line end': "Ctrl+K",
+              'editor/kill to line start': "Ctrl+U",
+              'editor/yank': 'Meta+Y',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: This one is `Alt+Y`.

File: spyderlib/config/main.py [hunk 19]
```diff
@@ -580,6 +580,51 @@ def is_ubuntu():
...
+              'editor/rotate kill ring': 'Shift+Ctrl+Y',
+              'editor/kill previous word': 'Ctrl+Backspace',
+              'editor/kill next word': 'Ctrl+Delete',
+              'editor/start of document': 'Alt+Left',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: This is `Alt+<`, not `Alt+Left` (arrow)

File: spyderlib/config/main.py [hunk 20]
```diff
@@ -580,6 +580,51 @@ def is_ubuntu():
...
+              'editor/end of document': 'Alt+Right',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: And this is `Alt+>`

File: spyderlib/config/main.py [hunk 21]
```diff
@@ -580,6 +580,51 @@ def is_ubuntu():
...
+              # -- In widgets/editor
+              'editor/inspect current object': 'Ctrl+X, Ctrl+I',
+              'editor/go to line': 'Alt+G+G',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: This is activated by pressing `Alt+G`, so we can remove the second `G` :-)

File: spyderlib/config/main.py [hunk 22]
```diff
@@ -580,6 +580,51 @@ def is_ubuntu():
...
+              'editor/go to previous file': 'Ctrl+Tab',
+              'editor/go to next file': 'Ctrl+Shift+Tab',
+              '_/file switcher': 'Ctrl+X, Ctrl+B',
+              # -- In spyder.py
+              'editor/find text': "Ctrl+S",
+              'editor/find next': "F3",
+              'editor/find previous': "Ctrl+R",
+              'editor/replace text': "Ctrl+X, Ctrl+H",
+              'editor/undo': 'Ctrl+_',
+              'editor/cut': 'Ctrl+W',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: This one is not working, it's closing all files (I don't know why :-s)

File: spyderlib/config/main.py [hunk 23]
```diff
@@ -580,6 +580,51 @@ def is_ubuntu():
...
+              'editor/copy': 'Alt+W',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: This one is not working either. It just writes a `w` over the selected text, instead of copying it.

File: spyderlib/config/main.py [hunk 24]
```diff
@@ -580,6 +580,51 @@ def is_ubuntu():
...
+              'editor/paste': 'Ctrl+Y',
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: Not working either. I think the problem is these shortcuts are also defined in the Editor context menu, and so are causing conflicts with the shortcuts present in Preferences.

File: spyderlib/config/main.py [hunk 25]
```diff
@@ -580,6 +580,51 @@ def is_ubuntu():
...
+              'editor/delete': 'Delete',
+              'editor/select all': "Ctrl+X+H",
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: Not working.

File: spyderlib/config/main.py [hunk 26]
```diff
@@ -580,6 +580,51 @@ def is_ubuntu():
...
+              # -- In plugins/editor
+              'editor/new file': "Ctrl+X, Ctrl+N",
+              'editor/open file': "Ctrl+X, Ctrl+F",
+              'editor/save file': "Ctrl+X, Ctrl+S",
```
Comment (ccordoba12) [inline spyderlib/config/main.py]: These double Ctrl shortcuts are also not working.

**Comment by ccordoba12** [comment]
A couple of comments:

1. There is a conflict with `Ctrl+Y`: It's assigned to both `redo` and `paste`. That's reported when you open Preferences.
2. None of the `Alt` shortcuts work unless you remove the `&` from all our main menus, for example `&File` (present in `spyder.py`). Although it's not required, I think it's better to remove `&` from submenus too.
3. After doing 2., i found that `Alt+F` and `Alt+B` (moving to next/prev word) are not working as expected. They fail to move among words when pressed repeatedly, i.e. they move only once to the next/prev word.
4. The UI in *Keyboard Shortcuts* should tell users what is the scheme they have currently selected.

----

It seems there's still some work to do to have this working correctly :-)

**Comment by blink1073** [comment]
The more I think about it, the more I think it is a mistake to offer this "mode".  It is too invasive and difficult to get right.  We shouldn't have to change the way the whole application works to support `Alt` keybindings, for instance. 

For example, "The UI in Keyboard Shortcuts should tell users what is the scheme they have currently selected."  This would no longer be strictly true once the user has customized the shortcuts.  

My preference would be to close this PR and leave everything as-is, where we support all of the required keybindings, and leave it up to the user to customize.

Thoughts @spyder-ide/core-developers?

**Comment by goanpeca** [comment]
@blink1073 do not despair (yet!). I think that letting the user know what bindings are on is beyond the point and can be left for another PR.

I will not use these keybidnings but if this will help more people use it I am willing to make the needed changes so that alt can work.

**Comment by blink1073** [comment]
I pushed emacs/redo to `Ctrl+Alt+Y`, since technically there is no redo in emacs.  Also, I fixed the prev/next word behavior.

**Comment by Nodd** [comment]
We could have multiple keybinds set, similar to how syntax highlighting works. There would be one tab for each set, with "Spyder", "Emacs", "Vim", "Custom" entries. Each set would be editable by the user (with a possibility to reset to defaults). The user the decides what set is active.
That may be too complicated to manage, but it's at elast coherent with syntax highlighting.

**Comment by goanpeca** [comment]
@Nodd I think we are moving away from that design in syntax highlighting, cause the defaults themes are already fixed. So the ides there was to have a single custom page where you could start from any of the shipped presets and then modify as you wished. Perhaps we could even add the save as a new custom, so you could choose among several customs in a single page.

But for shortcuts this page ideas.. might make sense.. but not sure :-s

**Comment by blink1073** [comment]
I am :-1: on multiple tabs.  I think it is confusing for syntax highlighting, and would be confusing here.

**Comment by ccordoba12** [comment]
I clearly understand what @Nodd means (always the voice of reason! ;-). However, as I see it, it only requires requires minor modifications to our current UI for Keyboard shortcuts.

I'll do a PR against @blink1073 PR to prove my point :-)

**Comment by Nodd** [comment]
Actually I'm not sure either if tabs are the correct solution.
Having different defaults to load (erasing all customization in the process) seems good enough and simpler for the ui, the user, the configuration file...

Looks like my voice of reason is a bit schizophrenic ! :wink:

**Comment by goanpeca** [comment]
Why not add a button to save keybindings that will pop out a dialog to put the name, and save it somewhere either in the ini or a separate set of files.

Also the user could export keybindings maybe?

**Comment by blink1073** [comment]
Both good points.

**Comment by blink1073** [comment]
Closing this one, as I don't have the bandwidth to work on it.

**Comment by ccordoba12** [comment]
@blink1073, please leave this one open. We know you can't work on this anymore, but we'll rebase your work and finish it for Spyder 3.3 ;-)

**Comment by jnsebgosselin** [comment]
http://doc.qt.io/qtcreator/creator-keyboard-shortcuts.html#emacs-shortcuts

**Comment by CAM-Gerlach** [comment]
@ccordoba12 Is this ever likely to get merged in its current form, or can it be closed for now?

**Comment by blink1073** [comment]
Closing this, since it is too far out of conflict.  New work would have to start from scratch.

**Comment by CAM-Gerlach** [comment]
Vim users rejoice :laughing: 

In all seriousness, its indeed unfortunate. However, hopefully in future versions we can provide better support for keyboard shortcut profiles...its something we've wanted to do for a long time, but we've always had more pressing priorities.

**Comment by Nodd** [comment]
Meanwhile vim users : https://github.com/spyder-ide/spyder-vim/issues/75 😭
